include/boost/geometry/algorithms/detail/buffer/turn_in_original_visitor.hpp

```diff
@@ -35,7 +35,12 @@ struct original_get_box
     template <typename Box, typename Original>
     static inline void apply(Box& total, Original const& original)
     {
-        geometry::expand(total, original.m_box);
+        typedef typename strategy::expand::services::default_strategy
+            <
+                box_tag, typename cs_tag<Box>::type
+            >::type expand_strategy_type;
```

Comment:
The CS should depend on the strategy passed by the user instead of being derived from the Box which AFAIU is based on the Point type of the geometry.

Reply:
The strategy passed (currently `IntersectionStrategy`) defines a segment strategy, and not a box or point strategy. So it is better to define it locally.

Besides that, I thought more about this and I don't see any point for a user-defined expand strategy... They are not earth-model dependant, the only special thing is that the dateline and the poles should be handled correctly. Do I miss something?

Reply:
> The strategy passed (currently IntersectionStrategy) defines a segment strategy, and not a box or point strategy.

It also defines the desired CS of the operation passed by the user. If you pass a strategy explicitly it defines how the operations are done. Geometries are used only at the highest level of an algorithm to get the default main strategy in order to pass it further. Internally the CS is defined always by the strategies.

> So it is better to define it locally. 

It's easier for sure. But it's wrong according to the design rationale of the library.

> I don't see any point for a user-defined expand strategy

Box, expanding, envelope, etc. are different in cartesian vs spherical/geographic. Just like any other CS-specific operation. Call e.g. `expand(BOX(170 0 179 1), POINT(-170 0))` and you'll see the difference which would be `BOX(-170 0 179 1)` in cartesian vs `BOX(170 0 190 1) in spherical/geographic`. Similar is in the case of a box expanded with another box. And I even not mentioned an area near poles where things are even wierder.

With that said this will work, the boxes will simply be bigger than they could be. However we considered passing cartesian geometries and geographic strategy. What if you do the opposite? You have geographic geometries and you pass cartesian strategy. Now this expand will take the winding of the CS at the antimeridian and the poles into account when the user wanted to calculate the cartesian intersection. Example:
1. User stores geographic geometries using `cs::geographic` containing e.g. `POINT(0, 50)`
2. User performs a map projection and stores the result using the same point type, now this point is e.g. `POINT(433000 526440)`
3. User calculates set operation or buffer passing the cartesian strategy explicitly (assuming that it is possible to pass a strategy into `buffer`).

At 3, internally at the line we're talking about here the CS is taken from the geometry and the geographic expand is calculated e.g. for this point `POINT(433000 526440)`. Technically this point is invalid geographical point. In practice `expand` will try to normalize them somehow. The result of this `expand` is going to be some arbitrary junk.

Reply:
Thanks for your answer. But this was not exactly my point. I know that expand should be different for cartesian and for spherical/geographic. But the CS is already part of the geometry, as you know. And that is exactly how I am deriving it.

I don't get the point about passing a cartesian geometry and a geographic strategy. Shouldn't that be avoided? But the way I implement it here there is no risk for that.

So my question is again: what is the need for a user-specific expand strategy? Will there ever be an Earth model relevant there?

Reply:
> I don't get the point about passing a cartesian geometry and a geographic strategy.

I always thought this was the intention. For example it was always possible to call:
```
distance(cart_pt1, cart_pt2, haversine<>());
```
and the result was correct and defined by the strategy.

It was also always possible to call `get_as_radian` with cartesian points.

> Shouldn't that be avoided?

The code compiles, the library doesn't complain about it, there is no static assert, there is no explicit information in the documentation about it AFAICS. You say "avoided" as if it was "it sometimes works and sometimes doesn't", it isn't a particularly firm and well-defined position. :)

I'm surprised that this issue is raised only now. After these pull requests adding the propagation of strategies and `cs::undefined` were merged:
https://github.com/boostorg/geometry/pull/533
https://github.com/boostorg/geometry/pull/605
So yes, it is possible to create geometries with undefined CS and then the user has to pass the strategy and the information about the CS is propagated downwards to all algorithms. As it is possible to not pass a strategy and use the default one taken from the geometry. So it works both ways where the strategy has the priority.

> But the way I implement it here there is no risk for that.

You mean this code will only be called in one CS?

> So my question is again: what is the need for a user-specific expand strategy? Will there ever be an Earth model relevant there?

It doesn't matter if we consider the earth model or the strategy containing it. The earth model also has the information about the CS. My reasoning is based on the fact that if the user passes a strategy, this strategy defines a CS/model and that all of the strategies used in an algorithm has to be consistent with each other.

At this point we have a few options:
0. continue without expressing explicitly what the behavior is when CSes of geometries and strategies are mixed
1. enforce a match of the CS between geometries and strategies with a static assert which I think is a breaking change at this point
2. only use the CS information of the geometry, impossible since strategies can be passed
3. only use the CS information of the strategy, impossible as well because of the default behavior is defined this way
4. allow both with one preferred behavior, define all CS-related information in the strategy with the geometry defining the default behavior.

I thought that 4. was always the intention but we were not pursuing it well enough and were conveniently avoiding it.

Btw, if the propagation of the model is the only problem here, why do we even need an umbrella strategy? If the CS is defined by the geometry and the only purpose of the strategy passed by the user is to define the model then we could simply take this model and create all default strategies below from it. But the model also defines the CS. We'd still have the same problem here, i.e. if the model was the ellipsoid should we use cartesian expand or geographic/spherical expand?

My reasoning is that either we define everything CS-related in a strategy or neither of it. Otherwise it's confusing which information is taken from where.

Reply:
Thanks, I'll come back to this later.